provider/core/src/key.rs

```diff
@@ -148,6 +151,33 @@ impl Default for DataKeyMetadata {
     }
 }
 
+#[derive(Copy, Clone)]
+enum DataKeyPathImpl {
+    // This string literal is wrapped in leading_tag!() and trailing_tag!() to make it detectable
+    // in a compiled binary.
+    Tagged(&'static str),
+    // For runtime-created data keys, we don't include the tags.
+    Plain(&'static str),
+}
+
+impl DataKeyPathImpl {
+    pub fn get(&self) -> &'static str {
+        match self {
+            DataKeyPathImpl::Plain(s) => s,
+            DataKeyPathImpl::Tagged(s) => unsafe {
+                // This becomes const in 1.64
+                unsafe {
+                    // Safe due to invariant that self.path is tagged correctly
```

Comment:
nit: probably should mention this invariant on DataKeyPathImpl too, even though it's private